src/peer.js

```diff
@@ -259,56 +259,97 @@ class Peer extends EventEmitter {
    * Call Rest API and get the list of peerIds assciated with API key.
    * @param {function} cb - The callback function that is called after XHR.
    */
-  listAllPeers(cb = function() {}) {
-    if (!this._checkOpenStatus()) {
-      return;
+  listAllPeers(cb) {
+    // for Promise API
+    if (typeof cb !== 'function') {
+      console.log('promise api');
+      return this._listAllPeersPromise();
     }
 
-    const self = this;
-    const http = new XMLHttpRequest();
+    console.log('callback api');
+    // for Callback API
+    this._listAllPeersPromise()
+      .then(cb)
+      .catch(err => {
+        switch (err.message) {
+          case 'OPEN_ERROR': {
+            // do nothing(backward compatible)
+            break;
+          }
+          case 'HTTP_ERROR': {
+            this._abort('server-error', 'Could not get peers from the server.');
 
-    const url = `${this.socket.signalingServerUrl}/api/apikeys/${
-      this.options.key
-    }/clients/`;
+            cb([]);
+            break;
+          }
+          case 'SERVER_ERROR': {
+            cb([]);
+            break;
+          }
+          case 'PERMISSION_ERROR': {
+            const err = new Error(
+              "It doesn't look like you have permission to list peers IDs. " +
+                'Please enable the SkyWay REST API on dashboard'
+            );
+            err.type = 'list-error';
+            logger.error(err);
+            this.emit(Peer.EVENTS.error.key, err);
+
+            cb([]);
+            break;
+          }
+        }
+      });
+  }
 
-    // If there's no ID we need to wait for one before trying to init socket.
-    http.open('get', url, true);
+  /**
+   * Call Rest API and get the list of peerIds assciated with API key with Promise.
+   * @return {Promise} Promise resolved with results of API call.
+   */
+  _listAllPeersPromise() {
+    console.log('open check');
+    if (!this._checkOpenStatus()) {
+      return Promise.reject(new Error('OPEN_ERROR'));
+    }
 
-    /* istanbul ignore next */
-    http.onerror = function() {
-      self._abort('server-error', 'Could not get peers from the server.');
-      cb([]);
-    };
+    console.log('return promise');
+    return new Promise((resolve, reject) => {
+      const http = new XMLHttpRequest();
+      const url = `${this.socket.signalingServerUrl}/api/apikeys/${
+        this.options.key
+      }/clients/`;
 
-    http.onreadystatechange = function() {
-      if (http.readyState !== 4) {
-        return;
-      }
+      // If there's no ID we need to wait for one before trying to init socket.
```

Comment:
@alanmshelly I forgot what this comment exactly means. Could you tell me `ID` in `If there's no ID`, if you remember?

IIRC, I thinks this comment means `If there's no Peer ID in the response from signaling server, then we need to wait for other peers to establish a connection to the signaling server`.